src/music/Instruments.ts

```diff
@@ -0,0 +1,90 @@
+import Instrument from "./Instrument.js";
+import { Envelopes, AdsrConfig } from "./Envelopes.js";
+import { Note, Notes } from "./Notes.js";
+import { SampleData } from "./Samples.js";
+
+type OscType = typeof OscillatorNode.prototype.type;
+
+export interface OscillatorConfig {
+    type: OscType; // either 'sine', 'triangle', 'square', or 'sawtooth'
+    detune?: number; // detune in cents
+}
+
+export class AdsrOscillatorInstrument extends Instrument {
+
+    type: OscType;
+    _detune: number;
+    env: AdsrConfig;
+
+    constructor(osc: OscillatorConfig, env: AdsrConfig, volume: number = 1) {
+        super(volume);
+        this.type = osc.type;
+        this._detune = 1;
+        this.detune = osc.detune || 0;
+        this.env = env;
+    }
+
+    set detune(n: number) {
+        this._detune = Math.pow(2, n / 1200);
+    }
+
+    scheduleNote(context: AudioContext, note: Note): AudioNode {
+        const freqs: number[] = Notes.detuneWithCoeff(note.note, this._detune);
+        // note ending frequencies are either where they started or at the endNote
+        const endfreqs: number[] = (note.endNote === undefined) ? freqs : Notes.detuneWithCoeff(note.endNote, this._detune);
+        // create envelope
+        const gain: GainNode = Envelopes.createAdsrEnvelope(context, note.start, note.duration,
+            this.env, Math.min(1 / freqs.length, 1) * this.volume);
```

Comment:
What is this volume scaling? Can you break this into an explanatory local variable or document what that's doing?